nnvm/python/nnvm/testing/check_computation.py

```diff
@@ -0,0 +1,451 @@
+
+    delta : float
+        A small value used for numerical gradient computation (usually called h in textbooks).
+
+    atol : float
+        Absolute tolerance.
+
+    rtol : float
+        Relative tolerance.
+
+    ng_atol : float
+        Absolute tolerance for numerical grad checking.
+
+    ng_rtol : float
+        Relative tolerance for numerical grad checking.
+
+    ng_max_error : float
+        Discard numerical partial derivatives whose estimated error is larger than this value.
+
+    ng_max_discarded_frac : float
+        Allow discarding no more than this fraction of partial derivatives.
+
+    dump_graph : bool
+        Dump the graph even on success.
+    """
+
+    if numerical_grads not in [False, True, 'if_possible']:
+        raise ValueError("numerical_grads must be a bool or 'if_possible', not {}"
+                         .format(numerical_grads))
+
+    input_vars = [x for x in symbol.list_input_variables()]
+    input_dict = {x.attr('name'): x for x in input_vars}
+
+    if grad_input_vars is None:
+        grad_input_vars = input_vars
+
+    shape = shape.copy() if shape else {}
+
+    grad_input_vars_real = []
+    for x in grad_input_vars:
+        if isinstance(x, tuple):
+            if len(x) != 2:
+                raise ValueError("Expected (var_name, shape), not {}".format(x))
+            x_name = x[0] if isinstance(x[0], str) else x[0].attr('name')
+            if x_name in shape:
+                if shape[x_name] != x[1]:
+                    raise ValueError("Can't decide which shape to use for {}: "
+                                     "{} from shape or {} from grad_input_vars"
+                                     .format(x_name, shape[x_name], x[1]))
+            shape[x_name] = x[1]
+            x = x[0]
+
+        grad_input_vars_real.append(input_dict[x] if isinstance(x, str) else x)
+
+    grad_input_vars = grad_input_vars_real
+
+    # Infer the output shape and dtype by creating a graph and running passes
+
+    forward_graph = nnvm.graph.create(symbol)
+
+    if dtype is not None:
+        nnvm.compiler.graph_attr.set_dtype_inputs(forward_graph, dtype)
+
+    if shape is not None:
+        nnvm.compiler.graph_attr.set_shape_inputs(forward_graph, shape)
+
+    forward_graph = forward_graph.apply('InferShape').apply('InferType')
+    shapes = forward_graph.json_attr('shape')
+    dtypes = forward_graph.json_attr('dtype')
+    out_shape = shapes[forward_graph.index.output_entries[0][0]]
+    out_dtype = dtypes[forward_graph.index.output_entries[0][0]]
+
+    backward_graph = None
+
+    # If we want gradients, we have to recreate the graph, but now with gradient computations
+    # Note that here we need out_shape for defining the shape of head grads, so we have to
+    # create the graph twice
+    if np_backward is not None or numerical_grads:
+        try:
+            head_grads_symbol = nnvm.symbol.Variable("head_grads", shape=out_shape, dtype=out_dtype)
```

Comment:
This function just supports computing grads of single output. Maybe we should support multiple outputs as well.